src/drstore.jl

```diff
@@ -0,0 +1,74 @@
+###
+# special struct to hold decision rules
+###
+
+# import Main
+struct DRStore{T}
+    data::Tuple{Vararg{Vector{T}}}
+    flat::Vector{T}
+end
+
+Main.length(ds::DRStore) = length(ds.data)
+Main.size(ds::DRStore, x...) = size(ds.data, x...)
+Main.getindex(ds::DRStore,x...) = getindex(ds.data,x...)
+Main.setindex!(ds::DRStore,x...) = setindex!(ds.data,x...)
+Main.abs(ds::DRStore{T}) where T = abs(ds.flat)
+
+Main.maxabs(ds::DRStore{T}) where T = maximum(e->norm(e,Inf), ds.flat)
+
+Main.norm(ds::DRStore{T}) where T = maximum(e->norm(e,Inf), ds.flat)
+# Main.norm(ds::DRStore{T},k::Int) where T = norm( (norm(e,k) for e in ds.flat), k)
+distance(ds1::DRStore{T}, ds2::DRStore{T}) where T = maximum( k->norm(k[1]-k[2],Inf), zip(ds1.flat,ds2.flat) )
+# distance(ds1::DRStore{T}, ds2::DRStore{T}, k::Int) where T = norm( (norm(e[1]-e[2],k) for e in zip(ds1.flat,ds2.flat)), k )
+
+total(ds::DRStore{T}) where T = sum(length(e) for e in ds.data)
+
+function DRStore(::Type{T}, dims::AbstractVector{Int}) where T
+    L = sum(dims)
+    fD = zeros(T, L)
+    ptr = pointer(fD)
+    sz = sizeof(T)
+    offsets = [[0]; cumsum(dims)[1:end-1]]*sz
+    D = [unsafe_wrap(Array{T}, ptr+offsets[i], dims[i]) for i=1:length(dims)]
+    return DRStore(tuple(D...), fD)
+end
+function DRStore(data::AbstractVector{S}) where S<:AbstractVector{T} where T
+    # data is always copied in this case
+    # as we cannot guarantee data is contiguous
+    dims = [length(d) for d in data]
+    ds = DRStore(T, dims)
+    for i=1:length(data)
+        ds.data[i][:] = data[i]
+    end
+    ds
+end
+
+# this one reuses data in place
+# should it be called DRStore! ?
+function DRStore(fD::Vector{T},dims::AbstractVector{Int}) where T
+    L = sum(dims)
+    ptr = pointer(fD)
+    sz = sizeof(T)
+    offsets = [[0]; cumsum(dims)[1:end-1]]*sz
+    D = [unsafe_wrap(Array{T}, ptr+offsets[i], dims[i]) for i=1:length(dims)]
```

Comment:
I think we should probably be testing that `fD` is the  correct length, right?

`unsafe_wrap` will let us have elements of `D` defined over memory that extends beyond `fD`:

```
julia> fD = rand(4)
4-element Array{Float64,1}:
 0.36243
 0.0790668
 0.232033
 0.975159


julia> X = unsafe_wrap(Array{Float64}, pointer(fD), 5)
5-element Array{Float64,1}:
 0.36243
 0.0790668
 0.232033
 0.975159
 9.88131e-324

julia> fD[:] = 1
1

julia> X
5-element Array{Float64,1}:
 1.0
 1.0
 1.0
 1.0
 9.88131e-324
```

(they _did_ warn us it was unsafe)

Reply:
I don't see anything wrong with the value `9.88131e-324`. I quite like it actually.
(joke aside: sure we need to check dimensions).